src/linear_algebra/bilinear_form.lean

```diff
@@ -1016,3 +1078,250 @@ begin
+
+variable [finite_dimensional K V]
+
+/-- Given a nondegenerate bilinear form `B`, `B.to_dual` is the isomorphism
+between a vector space and its dual with the underlying linear map `B.to_lin`. -/
+noncomputable def to_dual (B : bilin_form K V) (hB : B.nondegenerate) :
+  V ≃ₗ[K] module.dual K V :=
+B.to_lin.linear_equiv_of_ker_eq_bot
+  (nondegenerate_iff_ker_eq_bot.mp hB) subspace.dual_findim_eq.symm
+
+lemma to_dual_def {B : bilin_form K V} (hB : B.nondegenerate) {m n : V} :
+  B.to_dual hB m n = B m n := rfl
+
+end
+
+/-- Let `B` be a symmetric, nondegenerate bilinear form on a nontrivial module `M` over the ring
+  `R` with invertible `2`. Then, there exists some `x : M` such that `B x x ≠ 0`. -/
+lemma exists_bilin_form_self_neq_zero [htwo : invertible (2 : R)] [nontrivial M]
+  {B : bilin_form R M} (hB₁ : B.nondegenerate) (hB₂ : sym_bilin_form.is_sym B) :
+  ∃ x, ¬ B.is_ortho x x :=
+begin
+  by_contra, push_neg at h,
+  have : ∀ u v, 2 * B u v = 0,
+  { intros u v,
+    rw [show 2 * B u v = B u u + B v u + B u v + B v v,
+          by rw [is_ortho_def.1 (h u), is_ortho_def.1 (h v), hB₂ v u, two_mul, zero_add]; ring,
+        show B u u + B v u + B u v + B v v = B (u + v) (u + v),
+          by simp [← add_assoc], is_ortho_def.1 (h _)] },
+  have hcon : ∀ u v, B u v = 0,
+  { intros u v,
+    rw [show 0 = htwo.inv_of * (2 * B u v), by rw [this, mul_zero]], simp [← mul_assoc] },
+  exact let ⟨v, hv⟩ := exists_ne (0 : M) in hv $ hB₁ v (hcon v),
+end
+
+-- ↓ This lemma only applies in fields as we require `a * b = 0 → a = 0 ∨ b = 0`
+lemma span_singleton_inf_orthogonal_eq_bot
+  {B : bilin_form K V} {x : V} (hx : ¬ B.is_ortho x x) :
+  (K ∙ x) ⊓ B.orthogonal (K ∙ x) = ⊥ :=
+begin
+  rw ← finset.coe_singleton,
+  refine eq_bot_iff.2 (λ y h, _),
+  rcases mem_span_finset.1 h.1 with ⟨μ, rfl⟩,
+  have := h.2 x _,
+  { rw finset.sum_singleton at this ⊢,
+    suffices hμzero : μ x = 0,
+    { rw [hμzero, zero_smul, submodule.mem_bot] },
+    change B (μ x • x) x = 0 at this, rw [smul_left] at this,
+    exact or.elim (zero_eq_mul.mp this.symm) id (λ hfalse, false.elim $ hx hfalse) },
+  { rw submodule.mem_span; exact λ _ hp, hp $ finset.mem_singleton_self _ }
+end
+
+-- ↓ This lemma only applies in fields since we use the inverse
+lemma span_singleton_sup_orthogonal_eq_top {B : bilin_form K V}
+  (hB : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :
+  (K ∙ x) ⊔ B.orthogonal (K ∙ x) = ⊤ :=
+begin
+  refine eq_top_iff.2 (λ y _, _),
+  rw submodule.mem_sup,
+  refine ⟨(B x y * (B x x)⁻¹) • x, _, y - (B x y * (B x x)⁻¹) • x, _, _⟩,
+  { exact submodule.mem_span_singleton.2 ⟨(B x y * (B x x)⁻¹), rfl⟩ },
+  { intros z hz,
+    rcases submodule.mem_span_singleton.1 hz with ⟨μ, rfl⟩,
+    simp [is_ortho, mul_assoc, inv_mul_cancel hx, hB x] },
+  { simp }
+end
+
+lemma span_singleton_is_compl_orthogonal {B : bilin_form K V}
+  (hB : sym_bilin_form.is_sym B) {x : V} (hx : ¬ B.is_ortho x x) :
+  is_compl (K ∙ x) (B.orthogonal (K ∙ x)) :=
```

Comment:
You now have two copies of this lemma!  See `is_compl_span_singleton_orthogonal` below.

Reply:
Whoops, thanks